src/items/tile.cpp

```diff
@@ -365,10 +365,53 @@ void Tile::onAddTileItem(Item* item) {
 
 	if ((!hasFlag(TILESTATE_PROTECTIONZONE) || g_configManager().getBoolean(CLEAN_PROTECTION_ZONES))
 		&& item->isCleanable()) {
-		if (!dynamic_cast<HouseTile*>(this)) {
+		if (!this->getHouse()) {
 			g_game().addTileToClean(this);
 		}
 	}
+
+	if (item->isCarpet() && !item->isMoveable()) {
+		if (getTopTopItem() && getTopTopItem()->canReceiveAutoCarpet()) {
+			return;
+		}
+		auto house = getHouse();
+		if (!house) {
+			return;
+		}
+
+		std::vector<Tile*> surroundings = {
+			g_game().map.getTile(cylinderMapPos.x - 1, cylinderMapPos.y, cylinderMapPos.z),
+			g_game().map.getTile(cylinderMapPos.x + 1, cylinderMapPos.y, cylinderMapPos.z),
+			g_game().map.getTile(cylinderMapPos.x, cylinderMapPos.y - 1, cylinderMapPos.z),
+			g_game().map.getTile(cylinderMapPos.x, cylinderMapPos.y + 1, cylinderMapPos.z)
+		};
```

Comment:
NBC, this could be generated by a function and could be a vector of references or smart pointers.

Reply:
I didn't make this smart pointer because I'd have to customize the deleter in order to not have it free the Tile. Until we have the Tiles be smart pointers everywhere I think this would just make it dirtier.